### ISSUE TITLE
`scaled_dot_product_attention` silently ignore `attn_mask` with flash attention backend

### ISSUE THREAD

**Issue by shijie-wu** [pytorch/pytorch#107884]
### 🐛 Describe the bug

It seems that `torch.nn.functional.scaled_dot_product_attention` silently ignore `attn_mask` when calling the flash attention backend. Is my understanding correct? If so, could we produce a warning to prevent this?

https://github.com/pytorch/pytorch/blob/f6cce3c468284ded1137adb2520b56451b132f74/aten/src/ATen/native/transformers/attention.cpp#L681-L685

### Versions

main branch

cc @jbschlosser @bhosmer @cpuhrsch @erichan1 @drisspg

**Comment by drisspg** [comment]
I don't think what you claim is true. See:
https://github.com/pytorch/pytorch/blob/6c508e0be40a6b571a46949571b7d24de84a426c/test/test_transformers.py#L1353

https://github.com/pytorch/pytorch/blob/6c508e0be40a6b571a46949571b7d24de84a426c/aten/src/ATen/native/transformers/cuda/sdp_utils.cpp#L262

**Comment by shijie-wu** [comment]
Thanks for clarifying! Could this test be documented? Right now it's unclear under what condition each backend will be triggered unless one dig into the source code.

**Comment by drisspg** [comment]
@shijie-wu  There is description of explicit dispatch control here: https://pytorch.org/tutorials/intermediate/scaled_dot_product_attention_tutorial.html#explicit-dispatcher-control

If you attempt to run a backend that is not supported for your inputs and hardware it will raise a warning with the reasons why

**Comment by shijie-wu** [comment]
@drisspg thank you for the pointer! Could we link to this tutorial in https://pytorch.org/docs/stable/generated/torch.nn.functional.scaled_dot_product_attention.html?

**Comment by drisspg** [comment]
Sure, if you wanted to submit a PR I would be happy to review it!

**Comment by shijie-wu** [comment]
@drisspg thanks! https://github.com/pytorch/pytorch/pull/108772

**Comment by hypnopump** [comment]
But the problem of Flash Attention not having the Attention Mask option is still there, isn't it?

Flash Attn v2 introduced in [this PR](https://github.com/pytorch/pytorch/pull/105602) [does not support the `attn_mask` as input](https://github.com/pytorch/pytorch/blob/41e218984359a0dc93ee54776bba792b69b17b19/aten/src/ATen/native/transformers/attention.cpp#L721), unlike [the memory efficient version](https://github.com/pytorch/pytorch/blob/41e218984359a0dc93ee54776bba792b69b17b19/aten/src/ATen/native/transformers/attention.cpp#L737). 

Is there any plan to implement this option in Flash Attn?
